### PR TITLE
Add parsing-only support for typescript

### PR DESCRIPTION
**Pull request by anstarovoyt**
Added parsing support for typescript in script tags with ``lang="ts"``
```
<script lang="ts">
 export let test: number;
 </script>
```

There are some limitations of the implementation:
1. It uses WS internal TypeScript engine instead of TypeScript service integration
~~2. Auto-imports don't work (only imports by quick fixes) // I will take a look, most likely it can be easily fixed~~ fixed
3. It uses ``TypeScript`` as the content language unlike regular script tags that use an extension of js language ``SvelteJSLanguage``

### CODE REVIEW

File: src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt [hunk 1]
```diff
@@ -56,4 +64,15 @@ class SvelteHtmlLexer : HtmlLexer(InnerSvelteHtmlLexer(), false) {
     override fun isHtmlTagState(state: Int): Boolean {
         return state == _SvelteHtmlLexer.START_TAG_NAME || state == _SvelteHtmlLexer.END_TAG_NAME
     }
+
+    companion object {
+        fun getSvelteScriptContentProvider(mimeType: String?): HtmlScriptContentProvider? {
+            if (mimeType == "ts" || mimeType == "typescript") {
+                //todo here we should use language extension similar to SvelteJSLanguage, instead of the original lang
+                return LanguageHtmlScriptContentProvider.getScriptContentProvider(JavaScriptSupportLoader.TYPESCRIPT)
```
Comment (anstarovoyt) [inline src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: @tomblachut the main issue here is that we cannot directly add ``SvelteTypeScriptParser`` similar to ``SvelteJSParser`` because constructors of ``TypeScriptStatementParser`` and other classes are package-protected :(

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: I can try some solutions:
1. Add a delegate parser that extends ES6Parser
2. Add an inheritor inside ``com.intellij.lang.javascript.ecmascript6.parsing`` (but I am not sure that it will work because of IDE custom class loading)

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: @anstarovoyt I find 2nd solution really confusing for new contributors - IDK if it's something accepted in Java/Kotlin land, so my opinion is strictly first impression.

Delegate parser could be later maybe incorporated into WS (or you can change access modifier) so even if it's a bit harder to maintain for now it could be later refactored and removed from this repo.

We could also discuss whether custom langs are really necessary as long as stores and reactive statements are still WIP, currently I have no idea how to implement them without parser overrides

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: >We could also discuss whether custom langs are really necessary as long as stores and reactive statements are still WIP, currently I have no idea how to implement them without parser overrides

At this moment I see only one reason why we need a separate psi for such elements: "unresolved element" warning. But I think the problem can be solved using implicit elements, I will take a look. Do you see any other problems if we don't process reactive statements as real variable declarations?

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: if you don't see any issues with this PR, I think it is ready-to-merge as a basic ts support in svelte.

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: I finally had a chance to take a look at reactive declarations. First of all, I'm no longer sure #97 is even correct. Svelte has some edge cases around reactive declarations, they are not strict variable statements as far as I understand.

Now I'm looking into another approach, to cover them similarly to `SvelteJSReferenceExpression` from stores implementation, i.e. replacing `PsiElement` in `SvelteJSPsiBuilder`. `TypeScriptParser` ctor is public, so this would work without delegates/hacks etc.

`SvelteJSLabeledStatement.processDeclarations` needs to be overriden, but it's not all. Resolving from template to script needs some more code.
- `processDeclarations` in script tag calls
- `JSEmbeddedContentImpl.processDeclarations` calls
`JSResolveUtil.processDeclarationsInScope(this, processor, state, lastParent, place)`
- `JSScopeNamesCache.getNamedElementsInScope(scope, requiredName, false)`
- `ScopeNamesCache.JSNamesCacheVisitor` — it handles `visitJSVarStatement` but not `visitJSLabeledStatement`

My idea is to implement both `JSLabeledStatement` and `JSVarStatement` by `SvelteJSLabeledStatement` OR rework resolving to skip `JSEmbeddedContentImpl.processDeclarations` and directly process each statement in `JSEmbeddedContent`, not sure what is better right now.

Anyway, stores will work with patched `TypeScriptParser`, reactive declarations are subject to more investigation, so basic TS is within hands reach. I need to merge some PRs and publish a release this week, should've done it week ago already.

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: @tomblachut any updates?

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: @anstarovoyt You can take a peek at #163, I've recently took different approach to (initial) support for reactive declarations. It favours no inspections instead of confusing/wrong inspections and it does not require parser changes.

---

TS support will require couple of small additions on top of parser
* $$props etc resolving
* $storeSubscriptions - right now there is one way support, TS doesn't support it

SvelteJSReferenceExpressionResolver handles both of the above.

I think this will require 
```
public class SvelteTSLanguage extends JSLanguageDialect implements DependentLanguage {
    public static final SvelteTSLanguage INSTANCE = new SvelteTSLanguage();

    private SvelteTSLanguage() {
        super("SvelteTS", DialectOptionHolder.TS, JavaScriptSupportLoader.TYPESCRIPT);
    }
}
```
and mirrored `SvelteTSSpecificHandlersFactory` & `SvelteTSReferenceExpressionResolver`.  Same for `SvelteJSScriptContentProvider` & dedicated `IElementType`.

I hope that this SvelteTSLanguage won't break other things.

Having said that I' not sure what is best to do right now:
1. Merge this anyway and provide at least something for the users. They will be angry (at me) that above don't work.
2. First release next version with changes already in master, #146 & #163 and then finish missing pieces for TS.

Anyway I'm really happy that this reactive declarations vs TS problem is coming to an end.

Right now I'm going to sleep 😅

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: >Merge this anyway and provide at least something for the users. They will be angry (at me) that above don't work.
or
>First release next version with changes already in master, #163 and then finish missing pieces for TS.

I think I can fix the rest issues with the ts support in 1-2 days after merging #146 & #163 so we can do it all together

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: by the way. 
>$$props etc resolving
>$storeSubscriptions - right now there is one way support, TS doesn't support it
This PR doesn't inject TypeScript into html-specific things (only to the script tag), so everything should still work

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: Actually you can use $subscriptions in script tag also. 

I'm not sure how it works in terms of unwrapping the type, but for JS example please take a look at
https://svelte.dev/repl/130bc99a24d442afb27f0fd8f9ddec86?version=3.29.0

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: That's the biggest difference from Vue IMO, you need custom languages for everything, not only expressions

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: oh, I see. Thanks. I will try to fix it asap

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt:72]: Okay, thanks! I've merged #146 already and will merge #163 shortly.

Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt]: @tomblachut I've updated the PR

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteHtmlLexer.kt]: @anstarovoyt thanks I'll take a look in the evening

File: src/main/java/dev/blachut/svelte/lang/SvelteInspectionSuppressor.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import com.intellij.psi.PsiElement
+import dev.blachut.svelte.lang.psi.SvelteHtmlFile
+
+class SvelteInspectionSuppressor : InspectionSuppressor {
+    override fun isSuppressedFor(element: PsiElement, toolId: String): Boolean {
+        if (element.containingFile is SvelteHtmlFile) {
+            if (toolId == "UnnecessaryLabelJS") {
+                return element.textMatches("$")
+            }
+            if (toolId == "BadExpressionStatementJS") {
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/SvelteInspectionSuppressor.kt:14]: I didn't see problems from this inspection in JS, is this strictly TS related or did I miss something?

File: src/main/java/dev/blachut/svelte/lang/parsing/js/SvelteJSParser.kt [hunk 3]
```diff
@@ -22,4 +25,16 @@ class SvelteJSParser(builder: PsiBuilder) : ES6Parser<ES6ExpressionParser<*>, ES
             }
         }
     }
+
+    override fun buildTokenElement(type: IElementType) {
+        // there are too many places that uses element type JSElementTypes.REFERENCE_EXPRESSION,
+        // so use the new one only for the specific references
+        return super.buildTokenElement(
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/parsing/js/SvelteJSParser.kt:32]: Interesting, I tried to use `buildTokenElement` but had some problems with it, because not every `marker.done` goes through this method. 

I found one example that does not work that way:

https://svelte.dev/repl/65fc4b475b884dcba414139848ff02ef?version=3.29.0
```
const count = writable(0);
const x = { $count };
console.log(x); // Prints { $count: 0 }
```

Anyway this looks like weird side effect of stores implementation and not a feature so I guess it's okay to ignore it for now.

File: src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            return super.restrictByFileType(file, libraryService, moduleAndLibraryScope)
+                .uniteWith(GlobalSearchScope.getScopeRestrictedByFileTypes(moduleAndLibraryScope, file.fileType))
+        }
+    }
+
+    override fun getElementResolveScope(element: PsiElement): GlobalSearchScope? {
+        val psiFile = element.containingFile
+        if (psiFile?.fileType !is SvelteHtmlFileType) return null
+        if (psiFile !is XmlFile) return null
+        val scriptTag = JSUtils.findScriptTagContent(psiFile)
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt:28]: Svelte can have two script tags and they could have different languages, this finds first one, is it okay?

File: src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import com.intellij.lang.javascript.psi.resolve.JSElementResolveScopeProvider
+import com.intellij.lang.javascript.psi.util.JSUtils
+import com.intellij.lang.typescript.library.TypeScriptLibraryProvider
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.psi.xml.XmlFile
+import dev.blachut.svelte.lang.SvelteHtmlFileType
+
+class SvelteElementResolveScopeProvider : JSElementResolveScopeProvider {
```
Comment (tomblachut) [inline src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt:14]: THB I don't understand that this class does

**Review by tomblachut — APPROVED** [review]
Excellent, thanks so much. I left some notes, please respond in your spare time.

I'll go ahead and add completions for script lang attribute similar to those for style tag, and create next release.

#32 can be closed with this, right? LSP support is not really urgent now I think.

File: src/main/java/dev/blachut/svelte/lang/SvelteInspectionSuppressor.kt (hunk 2 above)
Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/SvelteInspectionSuppressor.kt:14]: it fixes useless warning for ``$: name``

File: src/main/java/dev/blachut/svelte/lang/parsing/js/SvelteJSParser.kt (hunk 3 above)
Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/js/SvelteJSParser.kt:32]: yes, I think so. Unfortunately, we cannot get the token text using the other implementation because it is called after "advance()" method

File: src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt (hunk 4 above)
Comment (anstarovoyt) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt:28]: I think we can live with it as is. I will try to find a better strategy but I am not sure

**Comment by anstarovoyt** [comment]
>#32 can be closed with this, right? LSP support is not really urgent now I think.

@tomblachut I am 100% sure that we still need to provide the integration with ts service but yes, it isn't a critical task for now.

**Comment by tomblachut** [comment]
Okay

File: src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt (hunk 4 above)
Comment (dummdidumm) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt:28]: We way we handle this in the language server is to check script and then module script and pick the first lang/type tag we find, so if someone chooses TS in the script tag he will automatically use it in the other one as well - no language mixes.

Comment (tomblachut) [reply src/main/java/dev/blachut/svelte/lang/parsing/html/SvelteElementResolveScopeProvider.kt:28]: Seems like a good balance 👍
